### DISCUSSION TITLE
UnboundLocalError when starting simulator server from a Thread

### DISCUSSION THREAD

**Discussion by elarandjelovic** [pymodbus-dev/pymodbus#2408] [Help]
I have a question relating to the dependency on aiohttp, which has come about as a result of trying to use the Pymodbus simulator in Pytest. I have a specific need to use the synchronous client, which unfortunately makes life difficult in Pytest because it means that my test is *not* in an asyncio context but the fixture which starts the simulation server is. To get around this I have created a fixture which starts up the simulation server in a separate Thread. This should work, but I came across an unexpected error in aiohttp triggered by the following code at the end of the ModbusSimulatorServer constructor:

```
app_key = getattr(web, 'AppKey', str)  # fall back to str for aiohttp < 3.9.0
self.api_key = app_key("modbus_server")
```

aiohttp throws an UnboundLocal error because it fails to find <module> in the frame stack when the server has been started in a new Thread.

If I downgrade aiohttp to a version pre 3.9.0 this doesn't happen and my test works fine.

Has anybody else come across this, and is there a way to avoid the problem without having to downgrade aiohttp? Many thanks in advance!

**Comment by janiversen** [comment]
Please have a look at our test suite, f.x. the sub_examples section, there we have a lot of tests with a sync client that connect to a async server in the same context.

**Comment by janiversen** [comment]
Our current version is 3.7.4, so how do you test with a version higher than 3.9.0

If you mean python version, then we currently test with 3.9 - 3.13, we no longer support 3.8, but it probably works.

**Reply by elarandjelovic** [reply]
Hi, thanks very much for the replies. I was referring to the version of aiohttp which is throwing the Exception in this scenario, as a result of this line:

[https://github.com/pymodbus-dev/pymodbus/blob/31528c1eafa020df7da55c751be1c843de92c33d/pymodbus/server/simulator/http_server.py#L220](https://github.com/pymodbus-dev/pymodbus/blob/31528c1eafa020df7da55c751be1c843de92c33d/pymodbus/server/simulator/http_server.py#L220)

I'll check out the test suite, thanks for the suggestion!
